pokemongo_bot/human_behaviour.py

```diff
@@ -1,46 +1,358 @@
+        lat_noise = gauss(0, radius/3.0)
+        lat_noise = min(max(-radius, lat_noise), radius)
+        
+        lng_noise = gauss(0, radius/3.0)
+        lng_noise = min(max(-radius, lng_noise), radius)
+
+        alt_noise = gauss(0, radius/3.0)
+        alt_noise = min(max(-radius, alt_noise), radius)
+    return lat_noise, lng_noise, alt_noise
+
+
+def ball_throw_hit_success_rng():
+    '''
+    Returns 1 if hits.
+
+    Configs:
+    "ball_throw_hit_success_chance": 1.0,
+
+    '''
+    ret = 0
+    if random() < get_config('ball_throw_hit_success_chance', 1.0):
+        ret = 1
+    return ret
+
+
+def aim_rng(target, std=0.05):
+    '''
+    noise from the target point should be approximated by gaussian,
+    we can wait for the missed reticles etc, which means we get
+    another chance to try...
+
+    TODO: add nice conf name for this...
+    
+    '''
+    for trial in range(10):
+        r = gauss(0, std)
+        if 0 <= target + r and target + r <= 1:
+            break
+    else:
+        # couldnt find target + gauss between [0, 1]
+        return random()
+    return target + r
 
 
 # Humanized `normalized_reticle_size` parameter for `catch_pokemon` API.
 # 1.0 => normal, 1.950 => excellent
-def normalized_reticle_size(factor):
-    minimum = 1.0
-    maximum = 1.950
-    return uniform(
-        minimum + (maximum - minimum) * factor,
-        maximum)
+def normalized_reticle_size_rng():
+    '''
+    Reticle rng.
+
+    Configs:
+    "replicate_ball_throw_reticle": "human",
+    "catch_randomize_reticle_factor": 0.9,
+
+    '''
+    mode = get_config('replicator.replicate_ball_throw_reticle', 'human')
+    factor = get_config('catch_randomize_reticle_factor', 0.9)
+
+    if 'exact' == mode:
+        return factor
+    elif 'uniform' == mode:
+        minimum = 1.0
+        maximum = 1.950
+        return uniform(
+            minimum + (maximum - minimum) * factor,
+            maximum)
+    elif 'human' == mode:
+        rnd = gauss(factor, 0.05)
+        # mirror the bounds
+        rnd = rnd%1.0
+        return 1.950 * rnd
+    return 1.950
 
 
 # Humanized `spin_modifier` parameter for `catch_pokemon` API.
 # 0.0 => normal ball, 1.0 => super spin curve ball
-def spin_modifier(factor):
-    minimum = 0.0
-    maximum = 1.0
-    return uniform(
-        minimum + (maximum - minimum) * factor,
-        maximum)
+def spin_modifier_rng():
+    '''
+    Spin rng.
+
+    Configs:
+    "replicate_ball_throw_spin" = "human",
+    "catch_randomize_spin_factor" = 0.9,
+
+    '''
+    mode = get_config('replicator.replicate_ball_throw_spin', 'human')
+    factor = get_config('catch_randomize_spin_factor', 0.9)
+
+    if 'exact' == mode:
+        return 1.0
+    elif 'uniform' == mode:
+        minimum = 0.0
+        maximum = 1.0
+        return uniform(
+            minimum + (maximum - minimum) * factor,
+            maximum)
+    elif 'human' == mode:
+        return aim_rng(factor)
+    return 1.0
+
+"""
+def _visualize():
```

Comment:
_visualize was used to draw graphs to check the distribution of rng (which I showed in #1980,) to verify that there are no anomalies at the bounds etc.
Suppose you use a gaussian of mu=0.5, with big std. If you clip the min and max value between [0, 1], then you will see pillars at 0 and 1, which is not nice as a rng to replicate human behaviour.
It might be useful for future rng development, however, the CI failed due to "import seaborn" line in the function (which is never used or called in the actual app, but detected by CI,) so I decided to comment them out rather than erasing.

The other function (_precalc_lognormal_ping_param) was used to calculate parameter for the log-normal. It uses theano, and since it is somewhat heavy task, I precalculated some of them and commented out for same reason, just incase we want to calculate parameter for our desired rng.